### PR TITLE
Detection task refactoring

### PR DESCRIPTION
**Pull request by jaegukhyun**
### Summary
This PR is for refactoring task of OTX Detection
I left old files(training_base.py nncf_base.py) for other task, and they will removed if we refactor all the other tasks.
New entrypoints:

-      base: otx.algorithms.detection.adapters.mmdet.task.MMDetectionTask
-      openvino: otx.algorithms.detection.adapters.openvino.task.OpenVINODetectionTask
-      nncf: otx.algorithms.detection.adapters.mmdet.nncf.task.DetectionNNCFTask

### Note
Adding new unit test will be handled in another PR due to this PR is too large already
Refactoring configurer will be handled in the next refactoring phase

### How to test
Normal unit-test, intg-test, e2e-test, regression test

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [x] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: otx/algorithms/detection/adapters/mmdet/configurer.py [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+        self.configure_data(cfg, training, data_cfg)
+        self.configure_regularization(cfg, training)
+        self.configure_task(cfg, training)
+        self.configure_hook(cfg)
+        self.configure_samples_per_gpu(cfg, subset)
+        self.configure_fp16_optimizer(cfg)
+        self.configure_compat_cfg(cfg)
+        return cfg
+
+    def base_configure(self, cfg, data_cfg, data_classes, model_classes):
```
Comment (sungmanc) [inline otx/algorithms/detection/adapters/mmdet/configurer.py]:
```suggestion
    def configure_base(self, cfg, data_cfg, data_classes, model_classes):
```

File: otx/algorithms/detection/adapters/mmdet/configurer.py [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+from otx.algorithms.detection.adapters.mmdet.utils import (
+    patch_datasets,
+    patch_evaluation,
+)
+
+logger = get_logger()
+
+
+# pylint: disable=too-many-public-methods
+class DetectionConfigManager:
```
Comment (sungmanc) [inline otx/algorithms/detection/adapters/mmdet/configurer.py]:
```suggestion
class DetectionConfigurer:
```

File: otx/algorithms/detection/adapters/mmdet/utils/exporter.py [hunk 3]
```diff
@@ -1,39 +1,35 @@
...
-from .stage import DetectionStage
+from otx.algorithms.detection.adapters.mmdet.utils.builder import build_detector
 
 logger = get_logger()
 
 
-@STAGES.register_module()
-class DetectionExporter(ExporterMixin, DetectionStage):
-    """Export class for object detection."""
+class DetectionExporter(Exporter):
```
Comment (sungmanc) [inline otx/algorithms/detection/adapters/mmdet/utils/exporter.py:18]: What is the differences between `MMDetectionTask.export()`?
As far as I remember, did we decide to delete this ?

**Review by sungmanc — COMMENTED** [review]
What a wonderful progress :) Could you check my comments?

File: otx/algorithms/detection/adapters/mmdet/configurer.py (hunk 1 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/adapters/mmdet/configurer.py]: It is base configurations step, not configure base.

File: otx/algorithms/detection/adapters/mmdet/configurer.py (hunk 2 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/adapters/mmdet/configurer.py]: If we use DetectionConfigManager, ClassficationConfigManager, etc, could this lead confusion with existed ConfigManger?

File: otx/algorithms/detection/adapters/mmdet/utils/exporter.py (hunk 3 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/adapters/mmdet/utils/exporter.py:18]: I want to bring this codes to export, but it's hard to bring previous exporter_mixin code. And I think exporter is already well implemented, so it is not necessary to bring to task.

File: otx/algorithms/detection/adapters/mmdet/configurer.py (hunk 1 above)
Comment (sungmanc) [reply otx/algorithms/detection/adapters/mmdet/configurer.py]: what is the difference between the two? I'm just wondering why this function use `base_configure` although all other configure functions used `configure_****` name.

```python 3
self.base_configure(cfg, data_cfg, data_classes, model_classes)
self.configure_device(cfg)
self.configure_model(cfg, ir_options)
self.configure_ckpt(cfg, model_ckpt)
self.configure_data(cfg, training, data_cfg)
self.configure_regularization(cfg, training)
self.configure_task(cfg, training)
self.configure_hook(cfg)
self.configure_samples_per_gpu(cfg, subset)
self.configure_fp16_optimizer(cfg)
self.configure_compat_cfg(cfg)
```

File: otx/algorithms/detection/adapters/mmdet/configurer.py (hunk 2 above)
Comment (sungmanc) [reply otx/algorithms/detection/adapters/mmdet/configurer.py]: That's the reason why I suggested to you to use `DetectionConfigurer`

File: otx/algorithms/detection/adapters/mmdet/utils/exporter.py (hunk 3 above)
Comment (sungmanc) [reply otx/algorithms/detection/adapters/mmdet/utils/exporter.py:18]: > I want to bring this codes to export, but it's hard to bring previous exporter_mixin code. And I think exporter is already well implemented, so it is not necessary to bring to task.

I agree with this point of view, however, coexisting of two similar function could make users to confuse. (`Detectiontask.export()`, `DetectionExporter`).

Then, do you agree with the suggestion that removes `DetectionExporter` in the future?

File: otx/algorithms/detection/adapters/mmdet/configurer.py (hunk 1 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/adapters/mmdet/configurer.py]: For example configure_device is configuring device setting, but base_configure do not contain exact object for configuration. Therefore I make this function as a base configuration function before configuration. However if you think this is not make sense, then I'll change of course. I want to explain my point

File: otx/algorithms/detection/adapters/mmdet/utils/exporter.py (hunk 3 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/adapters/mmdet/utils/exporter.py:18]: I hope, but I'm not sure

Comment (jaegukhyun) [reply otx/algorithms/detection/adapters/mmdet/utils/exporter.py:18]: And also, there are a lot of works for MMDetectionTask.export() except works for DetectionExporter

Comment (harimkang) [reply otx/algorithms/detection/adapters/mmdet/utils/exporter.py:18]: It looks like it needs to be refactored with mixin in the future. But it seems hard to do in this PR.

File: otx/algorithms/detection/adapters/mmdet/configurer.py (hunk 1 above)
Comment (harimkang) [reply otx/algorithms/detection/adapters/mmdet/configurer.py]: How about `configure_base`? or `configure_init`?

File: otx/algorithms/common/adapters/mmcv/utils/config_utils.py [hunk 4]
```diff
@@ -391,12 +392,152 @@ def patch_runner(config: Config):
...
+    """
+    if not torch.cuda.is_available() and "fp16" in config:
+        logger.info("Revert FP16 to FP32 on CPU device")
+        if isinstance(config, Config):
+            del config._cfg_dict["fp16"]  # pylint: disable=protected-access
+        elif isinstance(config, ConfigDict):
+            del config["fp16"]
+
+
+def update_basic_hooks(config: Config):
```
Comment (harimkang) [inline otx/algorithms/common/adapters/mmcv/utils/config_utils.py]: Splitting the two hooks added here into two functions seems like a good idea.
In fact, if it's going to be gone in the future, it's probably okay to leave it there.

**Review by harimkang — COMMENTED** [review]
LGTM, I'll ask you directly about things that come up while working on the classification task. In this PR, I made one comment.

File: otx/algorithms/detection/adapters/mmdet/utils/exporter.py (hunk 3 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/adapters/mmdet/utils/exporter.py:18]: I agree, I think the whole Exporter part should be moved to _export_model function(currently not existed).

**Review by harimkang — DISMISSED** [review]

**Review by sungmanc — DISMISSED** [review]

File: otx/algorithms/detection/adapters/mmdet/configurer.py (hunk 1 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/adapters/mmdet/configurer.py]: I changed it to configure_base

File: otx/algorithms/common/adapters/mmcv/utils/config_utils.py (hunk 4 above)
Comment (jaegukhyun) [reply otx/algorithms/common/adapters/mmcv/utils/config_utils.py]: I divided it into patch_adaptive_interval_training and patch_early_stopping

**Review by harimkang — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]

**Pull request merged**